arb_os/txReceipt.mini

```diff
@@ -0,0 +1,148 @@
+
+// This is declared identically in messages.mini
+type IncomingRequest = struct {
+    kind: uint,               // type of message
+    blockNumber: uint,        // block number of the L1 block
+    timestamp: uint,          // timestamp of the L1 block
+    sender: address,          // address of the sender
+    requestId: uint,          // sequence number in L1 inbox
+    msgData: MarshalledBytes  // contents of the message, as a marshalled bytearray
+}
+
+// This is declared identially in gasAccounting.mini
+type GasUsage = struct {
+    gasUsed: uint,
+    gasPriceWei: uint,
+}
+
+type TxResultInfo = struct {
+    returnCode: uint,
+    returnData: MarshalledBytes,
+    evmLogs: EvmLogs,
+    perBlockData: PerBlockReceiptData,
+}
+
+type PerBlockReceiptData = struct {
+    totalGasUsed: uint,
+    numTx: uint,
+    numEvmLogs: uint,
+}
+
+var globalBlockReceiptData: struct {
+    blockNum: uint,
+    data: PerBlockReceiptData
+};
+
+public impure func txReceipts_init() {  // will be called at beginning of main()
+    init_txReceiptsForBlock(0);
+}
+
+impure func init_txReceiptsForBlock(blockNum: uint) {
+    globalBlockReceiptData = struct {
+        blockNum: blockNum,
+        data: struct {
+            totalGasUsed: 0,
+            numTx: 0,
+            numEvmLogs: 0
+        }
+    };
+}
+
+impure func update_txReceiptsForBlock(
+    blockNum: uint,
+    gasUsed: uint,
+    numEvmLogs: uint
+) -> PerBlockReceiptData {
+    if (blockNum > globalBlockReceiptData.blockNum) {
+        init_txReceiptsForBlock(blockNum);
+    }
+
+    // first add in the gas, because returned value is supposed to include it
+    let ret = globalBlockReceiptData.data with {
+        totalGasUsed: gasUsed + globalBlockReceiptData.data.totalGasUsed
+    };
+
+    // now update the accumulated data and write it back
+    globalBlockReceiptData = globalBlockReceiptData with {
+        data: ret with {
+            numTx: ret.numTx + 1
+        } with {
+            numEvmLogs: ret.numEvmLogs + numEvmLogs
+        }
+    };
+
+    return ret;
+}
+
+// result codes:
+//    0    return (success)
+//    1    tx reverted
+//    2    rejected due to congestion
+//    3    insufficient funds to pay for gas
+//    4    insufficient funds for tx payment
+//    5    bad sequence number
+//    6    message format error
+//    255  unknown error
+public impure func emitLog(
+    l1message: IncomingRequest,
+    resultCode: uint,
+    maybeReturnData: option<ByteArray>,
+    maybeEvmLogs: option<EvmLogs>,
+    gasUsage: option<GasUsage>,
+) {
+    let returnData = bytearray_new(0);
+    if let Some(rd) = maybeReturnData {
+        returnData = rd;
+    }
+    let evmLogs = evmlogs_empty();
+    if let Some(el) = maybeEvmLogs {
+        evmLogs = el;
+    }
+    let realGasUsage = struct {
+        gasUsed: 0,
+        gasPriceWei: 0,
+    };
+    if let Some(gu) = gasUsage {
+        realGasUsage = gu;
+    }
+
+    let txResultInfo = struct {
```

Comment:
I got confused by this struct being called `txResultInfo` and the other `TxResultInfo` struct

Reply:
Resolved this by eliminating the top-level `TxResultInfo` type, which isn't needed anyway.